File: az_it/az_it/js_scripts/lead.js
// Client Script for Lead DocType
// Purpose: Cancel all associated ToDos when Lead status changes to "Lost Quotation" or "Do Not Contact"
// Event: after_save

frappe.ui.form.on('Lead', {
    after_save: function(frm) {
        // Check if status is one of the target statuses that should close ToDos
        if (frm.doc.status === 'Lost Quotation' || frm.doc.status === 'Do Not Contact') {
            
            // Call server method to cancel all associated ToDos
            frappe.call({
                method: 'az_it.az_it.python_scripts.lead.cancel_todos_for_lead',
                args: {
                    lead_id: frm.doc.name,
                    lead_status: frm.doc.status
                },
                callback: function(response) {
                    if (response.message && response.message.cancelled_count > 0) {
                        frappe.show_alert({
                            message: __('Cancelled {0} ToDo(s) associated with this Lead', [response.message.cancelled_count]),
                            indicator: 'green'
                        }, 5);
                    }
                },
                error: function(err) {
                    console.error('Error cancelling ToDos:', err);
                }
            });
        }
    }
});

File: az_it/az_it/python_scripts/lead.py

import frappe

@frappe.whitelist()
def cancel_todos_for_lead(lead_id, lead_status):

    # Verify that the status is one we should act on
    if lead_status not in ['Lost Quotation', 'Do Not Contact']:
        return {'cancelled_count': 0, 'message': 'No action needed for this status'}
    
    try:
        # Get all open ToDos for this Lead
        todos = frappe.db.get_all(
            'ToDo',
            filters={
                'reference_type': 'Lead',
                'reference_name': lead_id,
                'status': 'Open'
            },
            fields=['name', 'description']
        )
        
        cancelled_count = 0
        
        # Cancel each ToDo
        for todo in todos:
            todo_doc = frappe.get_doc('ToDo', todo.name)
            todo_doc.status = 'Cancelled'
            
            # Add a note about why it was cancelled
            if todo_doc.description:
                todo_doc.description += f"\n\n[Auto-cancelled: Lead status changed to '{lead_status}']"
            else:
                todo_doc.description = f"[Auto-cancelled: Lead status changed to '{lead_status}']"
            
            todo_doc.save()
            cancelled_count += 1
        
        # Commit the changes
        frappe.db.commit()
        
        
        return {
            'cancelled_count': cancelled_count,
            'message': f'Successfully cancelled {cancelled_count} ToDo(s)'
        }
        
    except Exception as e:
        frappe.throw(f"Error cancelling ToDos: {str(e)}")

